src/Microsoft.ML.Transforms/StatefulCustomMappingTransformer.cs

```diff
@@ -0,0 +1,398 @@
+        /// an input data with schema <paramref name="inputSchema"/>.
+        /// </summary>
+        public DataViewSchema GetOutputSchema(DataViewSchema inputSchema)
+        {
+            _host.CheckValue(inputSchema, nameof(inputSchema));
+            var rowToRow = new RowToRowMapper(_host, this, new EmptyDataView(_host, inputSchema));
+            return rowToRow.OutputSchema;
+        }
+
+        /// <summary>
+        /// Take the data in, make transformations, output the data.
+        /// Note that <see cref="IDataView"/>'s are lazy, so no actual transformations happen here, just schema validation.
+        /// </summary>
+        public IDataView Transform(IDataView input)
+        {
+            _host.CheckValue(input, nameof(input));
+            return new RowToRowMapper(_host, this, input);
+        }
+
+        /// <summary>
+        /// Constructs a row-to-row mapper based on an input schema. If <see cref="ITransformer.IsRowToRowMapper"/>
+        /// is <c>false</c>, then an exception is thrown. If the <paramref name="inputSchema"/> is in any way
+        /// unsuitable for constructing the mapper, an exception is likewise thrown.
+        /// </summary>
+        IRowToRowMapper ITransformer.GetRowToRowMapper(DataViewSchema inputSchema)
+        {
+            _host.CheckValue(inputSchema, nameof(inputSchema));
+            return new RowToRowMapper(_host, this, new EmptyDataView(_host, inputSchema));
+        }
+
+        private sealed class RowToRowMapper : RowToRowMapperTransformBase
+        {
+            private readonly StatefulCustomMappingTransformer<TSrc, TDst, TState> _parent;
+            private readonly ColumnBindings _bindings;
+            private readonly TypedCursorable<TSrc> _typedSrc;
+
+            public override DataViewSchema OutputSchema => _bindings.Schema;
+
+            public RowToRowMapper(IHostEnvironment env, StatefulCustomMappingTransformer<TSrc, TDst, TState> parent, IDataView input)
+                : base(env, "StatefulCustom", input)
+            {
+                Host.CheckValue(parent, nameof(parent));
+
+                _parent = parent;
+
+                var dstRow = new DataViewConstructionUtils.InputRow<TDst>(Host, _parent.AddedSchema);
+                // All the output columns of dstRow are our outputs.
+                var cols = Enumerable.Range(0, dstRow.Schema.Count).Select(x => new DataViewSchema.DetachedColumn(dstRow.Schema[x])).ToArray();
+
+                _bindings = new ColumnBindings(input.Schema, cols);
+
+                _typedSrc = TypedCursorable<TSrc>.Create(Host, input, false, _parent.InputSchemaDefinition);
+            }
+
+            public override DataViewRowCursor[] GetRowCursorSet(IEnumerable<DataViewSchema.Column> columnsNeeded, int n, Random rand = null)
+            {
+                return new[] { GetRowCursorCore(columnsNeeded, rand) };
+            }
```

Comment:
So I have a question.

My impression, is that this code in here is like this so to avoid this transform to use more than one thread, is that right?

Still, on the `ShouldUseParallelCursors` it returns null if the predicate there only wants columns that were already on the input of this transformer. 

So my question is, shouldn't this method here actually return multiple cursors if `ShouldUseParallelCursors` returns null??

To be honest, I don't know if what I am asking makes sense, but for instance there's the case of `GenerateNumberTransform`, where its `GetRowCursorSet` actually checks the `ShouldUseParallelCursors` to know if it returns 1 or more cursors.

https://github.com/dotnet/machinelearning/blob/4b3cc7e2253ac195f01533459f2fc8e98d3b0cea/src/Microsoft.ML.Data/Transforms/GenerateNumberTransform.cs#L361-L378

Reply:
Yes, you're right. `ShouldUseParallelCursors` is used in the base class `TransformBase`, inside the `IDataView.GetRowCursor` method. This method gets the cursor recursively from the `Source` of each `IDataTransform`, and in order to know whether to do that using `GetRowCursor` or `GetRowCursorSet` it calls `ShouldUseParallelCursors`. If this transform (`StatefulCustomMappingTransformer.RowToRowMapper`) is somewhere in the middle of the pipeline, and none of the transform's added columns are active, then it can safely produce multiple cursors from its `Source`.

---
In reply to: [401983989](https://github.com/dotnet/machinelearning/pull/4676#discussion_r401983989) [](ancestors = 401983989)